File: backend/app/schemas/organisation.py
import uuid
from datetime import datetime

from pydantic import BaseModel

from app.models.organisation import OrganisationRole


class OrganisationCreate(BaseModel):
    """Schema for creating a new organisation."""

    name: str


class OrganisationUpdate(BaseModel):
    """Schema for updating an organisation."""

    name: str | None = None


class OrganisationResponse(BaseModel):
    """Schema for organisation response."""

    id: uuid.UUID
    name: str
    is_personal: bool = False
    created_at: datetime

    model_config = {"from_attributes": True}


class OrganisationMemberResponse(BaseModel):
    """Schema for organisation member response."""

    user_id: uuid.UUID
    organisation_id: uuid.UUID
    role: OrganisationRole
    user_email: str | None = None
    user_name: str

    model_config = {"from_attributes": True}


class OrganisationWithRole(BaseModel):
    """Schema for organisation with the user's role."""

    id: uuid.UUID
    name: str
    role: OrganisationRole
    is_personal: bool = False
    created_at: datetime

    model_config = {"from_attributes": True}


class AddMemberRequest(BaseModel):
    """Schema for adding a member to an organisation."""

    user_id: uuid.UUID
    role: OrganisationRole = OrganisationRole.MEMBER


class UpdateMemberRoleRequest(BaseModel):
    """Schema for updating a member's role."""

    role: OrganisationRole

File: frontend/rooster_app/lib/models/organisation.dart
class Organisation {
  final String id;
  final String name;
  final String role; // "admin" or "member"
  final bool isPersonal;
  final DateTime createdAt;

  Organisation({
    required this.id,
    required this.name,
    required this.role,
    this.isPersonal = false,
    required this.createdAt,
  });

  bool get isAdmin => role == 'admin';

  factory Organisation.fromJson(Map<String, dynamic> json) {
    return Organisation(
      id: json['id'],
      name: json['name'],
      role: json['role'] ?? 'member',
      isPersonal: json['is_personal'] ?? false,
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'role': role,
      'is_personal': isPersonal,
      'created_at': createdAt.toIso8601String(),
    };
  }
}

class OrganisationMember {
  final String userId;
  final String organisationId;
  final String role;
  final String? userEmail;
  final String userName;

  OrganisationMember({
    required this.userId,
    required this.organisationId,
    required this.role,
    this.userEmail,
    required this.userName,
  });

  bool get isAdmin => role == 'admin';

  factory OrganisationMember.fromJson(Map<String, dynamic> json) {
    return OrganisationMember(
      userId: json['user_id'],
      organisationId: json['organisation_id'],
      role: json['role'] ?? 'member',
      userEmail: json['user_email'],
      userName: json['user_name'],
    );
  }
}
